### PR TITLE
[zlib-ng] Adapt recipe to be cross-compatible with Conan v1 and v2

### PR DESCRIPTION
**Pull request by jcar87**
Specify library name and version:  **zlib-ng/all**

Adapt recipe to be cross-compatible with v1 and v2. 
Tested with Conan 1.51.0 and Conan 2.0.0-beta2

### CODE REVIEW

File: recipes/zlib-ng/all/conanfile.py [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
+from conan.errors import ConanInvalidConfiguration
+from conan.tools import files
+from conan.tools.cmake import CMake, CMakeToolchain
 from conan.tools.microsoft import is_msvc
+from conan.tools.scm import Version
 import os
-import functools
 
-equired_conan_version = ">=1.33.0"
+equired_conan_version = ">=1.50.0"
```
Comment (jgsogo) [inline recipes/zlib-ng/all/conanfile.py:9]:
```suggestion
required_conan_version = ">=1.50.0"
```

**Comment by jgsogo** [comment]
Requires https://github.com/conan-io/conan-center-index/pull/11987

File: recipes/zlib-ng/all/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -1,9 +1,8 @@
-cmake_minimum_required(VERSION 3.1)
+cmake_minimum_required(VERSION 3.15)
```
Comment (SpaceIm) [inline recipes/zlib-ng/all/test_package/CMakeLists.txt:1]:
```suggestion
cmake_minimum_required(VERSION 3.1)
```

File: recipes/zlib-ng/all/conanfile.py [hunk 3]
```diff
@@ -51,35 +46,34 @@ def configure(self):
         del self.settings.compiler.cppstd
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version],
+        files.get(self, **self.conan_data["sources"][self.version],
             destination=self._source_subfolder, strip_root=True)
 
     def validate(self):
-        if self.options.zlib_compat and not self.options.with_gzfileop:
+        if self.info.options.zlib_compat and not self.info.options.with_gzfileop:
```
Comment (SpaceIm) [inline recipes/zlib-ng/all/conanfile.py:53]: is it valid even in conan v1 mode?

**Review by czoido — APPROVED** [review]

File: recipes/zlib-ng/all/conanfile.py [hunk 4]
```diff
@@ -51,35 +46,34 @@ def configure(self):
         del self.settings.compiler.cppstd
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version],
+        files.get(self, **self.conan_data["sources"][self.version],
             destination=self._source_subfolder, strip_root=True)
```
Comment (SpaceIm) [inline recipes/zlib-ng/all/conanfile.py:50]: would be better to define a layout, then refer to self.source_folder afterwards

File: recipes/zlib-ng/all/test_package/conanfile.py [hunk 5]
```diff
@@ -1,17 +1,26 @@
...
+
+from conan import ConanFile
+from conan.tools.cmake import CMake, cmake_layout
+from conan.tools.build import cross_building
 
 
 class TestPackageConan(ConanFile):
     settings = "os", "compiler", "build_type", "arch"
-    generators = "cmake", "cmake_find_package_multi"
+    generators = "CMakeDeps", "CMakeToolchain"
```
Comment (SpaceIm) [inline recipes/zlib-ng/all/test_package/conanfile.py:10]:
```suggestion
    generators = "CMakeDeps", "CMakeToolchain", "VirtualRunEnv"
```

Comment (jcar87) [reply recipes/zlib-ng/all/test_package/conanfile.py:10]: Thanks! Fixed.

**Review by prince-chrismc — APPROVED** [review]

File: recipes/zlib-ng/all/conanfile.py (hunk 3 above)
Comment (uilianries) [reply recipes/zlib-ng/all/conanfile.py:53]: Yes, it does. The only case I still have no idea how fix, is when a package can be header-only or not. When the package id sanitized, self.info.options.xxx and self.info.setting.xxx are None, so you can't validate and there is no get_safe available. I've opened an issue for more information: https://github.com/conan-io/conan/issues/11786

Comment (SSE4) [reply recipes/zlib-ng/all/conanfile.py:53]: I was told there will be new method `validate_build` to address that (doesn't look promising...)
right now
this works in v2, but fails in v1:
```
cppstd = self.info.settings.get_safe("compiler.cppstd")
```
this works in v1, but fails in v2:
```
cppstd = self.settings.get_safe("compiler.cppstd")
```
the only way I have found to satisfy both:
```
if self.info.settings.get_safe is not None:
    cppstd = self.info.settings.get_safe("compiler.cppstd")
else:
    cppstd = self.settings.get_safe("compiler.cppstd")
```
(even `hasattr` didn't work as expected 🤔 )

**Review by SSE4 — APPROVED** [review]

File: recipes/zlib-ng/all/conanfile.py (hunk 3 above)
Comment (lasote) [reply recipes/zlib-ng/all/conanfile.py:53]: This recipe is correct about the `validate` as it is checking if a binary is possible or not.
I responded to the issue opened by @uilianries, I hope now it makes sense: https://github.com/conan-io/conan/issues/11786

Comment (lasote) [reply recipes/zlib-ng/all/conanfile.py:53]: @SSE4 Please let us know, in the conan developer channel why you consider this feature `doesn't look promising` or open an issue as @uilianries did. I don't think that sentence is the optimal way to express doubts about a feature that your team designed.

**Pull request merged**